feat: implement more complex config handling

alexander-jackson opened the pull request:
Firstly, change job configurations to expect `message_creation_timestamp` and `prediction_cutoff_timestamp` fields, which inform the client of when the DCL processed the message and when it expects predictions to be returned by. Sending the creation time gives a better idea of the latency between the two and allows the client to calculate the estimated time of the job. Sending the cutoff time provides a fixed point in time at which the client should return by, meaning they can stop early if processing is taking longer than expected without being penalised.

Additionally, allow users to specify a custom function for handling job configurations. This allows for more advanced usage of the data provided to them, such as using different timeouts for different job types as shown in the unit tests.

alexander-jackson committed:
feat: implement more complex config handling

Firstly, change job configurations to expect
`message_creation_timestamp` and `prediction_cutoff_timestamp` fields,
which inform the client of when the DCL processed the message and when
it expects predictions to be returned by. Sending the creation time
gives a better idea of the latency between the two and allows the client
to calculate the estimated time of the job. Sending the cutoff time
provides a fixed point in time at which the client should return by,
meaning they can stop early if processing is taking longer than expected
without being penalised.

Additionally, allow users to specify a custom function for handling job
configurations. This allows for more advanced usage of the data provided
to them, such as using different timeouts for different job types as
shown in the unit tests.

G-Kemp101 commented on sybl/client/job_config.py:
```diff
@@ -17,6 +19,21 @@ def __init__(
+        Args:
+            comparison_function: The function to use for comparing
+        """
+        self.comparison_function = comparison_function
```
Can you add a check here to verify the provided comparison_function is `callable` just so it is less likely to crash during the connection

alexander-jackson committed:
fix: check the comparison function is callable

Check that the comparison function that the client provides is actually
callable to save issues later.
